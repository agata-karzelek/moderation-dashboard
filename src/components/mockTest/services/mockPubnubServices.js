import { mockAccounts } from "../mockAccounts";
import { mockPermissions } from "../mockPermissions";
import { mockPubNubApplications } from "../mockPubnubAccounts";

const commonkeyValuePairs = {
  status: 200,
  statusText: "OK",
  headers: {
    "cache-control": "no-cache, private",
    "content-type": "application/json",
    expires: "Thu, 29 Jul 2021 15:37:05 GMT",
    pragma: "private",
  },
  config: {
    url: "/apps?owner_id=572238",
    method: "get",
    headers: {
      Accept: "application/json, text/plain, */*",
      "X-Session-Token": "rAL18qRqj8rhtxkJwzQ2dmRi5UrhH-vFbqUaLknxshs",
    },
    baseURL: "https://admin.pubnub.com/api",
    transformRequest: [null],
    transformResponse: [null],
    timeout: 0,
    xsrfCookieName: "XSRF-TOKEN",
    xsrfHeaderName: "X-XSRF-TOKEN",
    maxContentLength: -1,
    maxBodyLength: -1,
  },
  request: {},
};

export const mockAllAccounts = {
  data: {
    result: {
      accounts: mockAccounts,
      permissions: mockPermissions,
    },
  },
  ...commonkeyValuePairs,
};

export const mockAllApps = {
  data: mockPubNubApplications,
  ...commonkeyValuePairs,
};

export default null;
