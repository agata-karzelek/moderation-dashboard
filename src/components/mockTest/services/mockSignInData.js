import { mockPermissions } from "../mockPermissions";

export const mockSignInData = {
  data: {
    result: {
      created: 1627276598,
      expires: 1627634943,
      modified: 1627548543,
      role: "user",
      status: 1,
      storage: null,
      token: "dGgB0-lLGG5GmjFgtr6t2nBFqgM05Nijq_YzzYdxVUk",
      user: {
        created: 1619760527,
        deleted: 0,
        email: "sourcefusepubnub@gmail.com",
        id: 572287,
        login_attempts: 0,
      },
      user_id: 572287,
      user_roles: {
        is_internal_admin: false,
        is_internal_operator: false,
        is_internal_viewer: false,
        roles: mockPermissions,
      },
    },
  },
  status: 200,
  statusText: "OK",
};

export default null;
