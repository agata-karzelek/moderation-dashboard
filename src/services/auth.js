import axios from "../utils/axios";

export async function signIn(credentials) {
  const signInResponse = await axios.post("/me", credentials);
  if (signInResponse.status === 200) {
    return signInResponse.data;
  }
  throw new Error(signInResponse.data.error || "Something went wrong");
}
