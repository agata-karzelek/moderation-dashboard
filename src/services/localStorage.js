import cookie from "js-cookie";

export const setCookie = (key, value) => {
  cookie.set(key, value, {
    expires: 1,
  });
};

export const removeCookie = (key) => {
  cookie.remove(key, {
    expires: 1,
  });
};

export const getCookie = (key) => {
  return cookie.get(key);
};

export const setLocalStorage = (key, value) => {
  localStorage.setItem(key, JSON.stringify(value));
};

export const removeLocalStorage = (key) => {
  localStorage.removeItem(key);
};

export const authenticate = (data, next) => {
  setCookie("token", data.token);
  setLocalStorage("user", data.user);
  next();
};

export const signout = (next) => {
  localStorage.clear();
  next();
};

export const isAuth = () => {
  const cookieChecked = getCookie("token");
  if (cookieChecked) {
    if (localStorage.getItem("user")) {
      return JSON.parse(localStorage.getItem("user"));
    } else {
      return false;
    }
  }
};
