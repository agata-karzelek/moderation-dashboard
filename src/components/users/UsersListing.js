import React, { useState } from "react";
import Helmet from "react-helmet";
import { Grid, Typography, Box } from "@material-ui/core";
import { formatDate, showError, usersFromLS } from "../../utils/helpers";
import Search from "../core/Search";
import UsersTable from "./UsersTable";
import AddUserMetadataModal from "./AddUserMetadataModal";
import usePubNub from "../../utils/usePubNub";
import { deleteUserMetadata, getUserByName } from "../../services/pubnub";
import FilterUsers from "./FilterUsers";
import SnackBar from "../core/SnackBar";

function UsersListing() {
  const [searchResult, setSearchResult] = useState([]);
  const [dataAdded, setDataAdded] = useState([]);
  const { pubnub } = usePubNub();
  const [searched] = useState("");
  const [filterBy, setFilterBy] = useState("");
  const [userAlert, setUserAlert] = useState({
    success: { status: false, msg: "" },
    error: { status: false, msg: "" },
  });

  const requestUserSearch = (searchedVal) => {
    setUserAlert({
      ...userAlert,
      success: { status: false, msg: "" },
      error: { status: false, msg: "" },
    });
    (async () => {
      setUserAlert({
        ...userAlert,
        success: { status: false, msg: "" },
        error: { status: false, msg: "" },
      });
      try {
        const filteredRows = await getUserByName(pubnub, searchedVal);
        const usersList = [];
        let selectedUser = {};
        if (!filteredRows.length) {
          setUserAlert({
            ...userAlert,
            success: { status: false, msg: "" },
            error: { status: true, msg: "No users found" },
          });
        }
        filteredRows.map((user, index) => {
          selectedUser = user;
          selectedUser.updated = formatDate(user.updated);
          usersList.push(selectedUser);
          return false;
        });
        setSearchResult(usersList);
      } catch (e) {
        setUserAlert({
          ...userAlert,
          success: { status: false, msg: "" },
          error: { status: true, msg: "Failed to filter user" },
        });
      }
    })();
  };
  const cancelUserSearch = () => {
    setSearchResult(usersFromLS());
  };

  const setSearchableData = (usersData) => {
    setSearchResult(usersData);
  };

  const deleteUser = (UUID) => {
    if (UUID) {
      (async () => {
        try {
          await deleteUserMetadata(pubnub, UUID);
          setDataAdded([]);
          setUserAlert({
            ...userAlert,
            success: { status: true, msg: "User deleted successfully" },
            error: { status: false, msg: "" },
          });
        } catch (e) {
          setUserAlert({
            ...userAlert,
            success: { status: false, msg: "" },
            error: { status: true, msg: showError(e.status.errorData) },
          });
        }
      })();
    }
  };

  const userAdded = (data) => {
    setDataAdded(data);
  };

  const handleFilterChange = (value) => {
    let filter = "";
    if (value !== "all") {
      filter = `custom.${value} == true`;
    }
    setFilterBy(filter);
  };

  return (
    <>
      <Helmet title="Users" />
      <Grid justify="space-between" container>
        <Grid item sm={4}>
          <Box pl={1}>
            <Typography testid="title" variant="h5">
              Users
            </Typography>
          </Box>
        </Grid>
        <Grid item sm={8}>
          <Box display="flex" justifyContent="flex-end">
            <Box ml={2}>
              <Search
                searched={searched}
                requestSearch={requestUserSearch}
                cancelSearch={cancelUserSearch}
                placeholder={"Search for user name"}
              />
            </Box>
            <Box ml={2}>
              <FilterUsers isFiltered={handleFilterChange} />
            </Box>
            <Box ml={2}>
              <AddUserMetadataModal pubnubObject={pubnub} isAdded={userAdded} />
            </Box>
          </Box>
        </Grid>
      </Grid>
      <br />
      <Grid container spacing={6}>
        <Grid item xs={12}>
          <UsersTable
            searchResult={searchResult}
            setSearchableData={setSearchableData}
            deleteUser={deleteUser}
            data={dataAdded}
            pubnub={pubnub}
            filterBy={filterBy}
          />
        </Grid>
      </Grid>
      {userAlert.error.status && <SnackBar msg={userAlert.error.msg} status={"info"} />}
      {userAlert.success.status && <SnackBar msg={userAlert.success.msg} status={"success"} />}
    </>
  );
}

export default UsersListing;
