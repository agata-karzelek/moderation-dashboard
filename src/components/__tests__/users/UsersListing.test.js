import React from "react";
import { shallow, mount } from "enzyme";
import UsersListing from "../../users/UsersListing";
import { act } from "react-dom/test-utils";
import { deleteUserMetadata, getUserByName } from "../../../services/pubnub";
jest.mock("../../../services/pubnub", () => {
  return {
    deleteUserMetadata: jest.fn(),
    getUserByName: jest.fn(),
  };
});
const waitForComponentToPaint = async (wrapper) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve));
    wrapper.update();
  });
};
const failPromise = {
  status: {
    errorData: {
      message: "Forbidden",
    },
  },
};
describe("User List Test Cases", () => {
  afterAll(() => {
    jest.resetAllMocks();
  });
  test("User List Header", () => {
    const wrapper = shallow(<UsersListing />);
    expect(wrapper.find("[testid='title']").text()).toBe("Users");
  });

  test("User Snapshot", () => {
    const wrapper = shallow(<UsersListing />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Have Header Tools and User Table", () => {
    const wrapper = shallow(<UsersListing />);
    expect(wrapper.find("Search").exists()).toBe(true);
    expect(wrapper.find("FilterUsers").exists()).toBe(true);
    expect(wrapper.find("AddUserMetadataModal").exists()).toBe(true);
    expect(wrapper.find("UsersTable").exists()).toBe(true);
  });

  test("Search Box Input Change", async () => {
    const wrapper = mount(<UsersListing />);

    // before the state updated
    await waitForComponentToPaint(wrapper);
    // after the state updated

    const searchBox = wrapper.find("input[type='text']").getElement().props;
    expect(searchBox.value).toBe("");
    expect(searchBox.placeholder).toBe("Search for user name");

    wrapper.find("input[type='text']").simulate("change", {
      target: { value: "newText" },
    });
    const newSearchBox = wrapper.find("input[type='text']").getElement().props;
    expect(newSearchBox.value).toBe("newText");
  });

  test("delete user event", async () => {
    const wrapper = mount(<UsersListing />);
    // before the state updated
    await waitForComponentToPaint(wrapper);
    await act(async () => {
      // filter user
      expect(wrapper.find("FilterUsers").getElements()[0].props.isFiltered("all")).toBeUndefined();
      expect(wrapper.find("FilterUsers").getElements()[0].props.isFiltered("one")).toBeUndefined();
      // added user
      expect(wrapper.find("AddUserMetadataModal").getElements()[0].props.isAdded()).toBeUndefined();
      // delete user
      wrapper.find("UsersTable").getElements()[0].props.deleteUser("1");
      expect(deleteUserMetadata).toHaveBeenCalledTimes(1);
      wrapper.find("UsersTable").getElements()[0].props.deleteUser();
      // error case for delete user
      deleteUserMetadata.mockImplementation(() => Promise.reject(failPromise));
      wrapper.find("UsersTable").getElements()[0].props.deleteUser("1");
      expect(deleteUserMetadata).toHaveBeenCalledTimes(2);

      // reset and cancel search
      expect(getUserByName).toHaveBeenCalledTimes(0);
      wrapper.find("Search").getElements()[0].props.requestSearch();
      expect(getUserByName).toHaveBeenCalledTimes(1);
      getUserByName.mockImplementation(() => [{ update: "15/08/2021" }]);
      wrapper.find("Search").getElements()[0].props.requestSearch();
      expect(getUserByName).toHaveBeenCalledTimes(2);
      wrapper.find("Search").getElements()[0].props.cancelSearch();

      // fail searching
      getUserByName.mockImplementation(() => failPromise);
      wrapper.find("Search").getElements()[0].props.requestSearch();
      expect(getUserByName).toHaveBeenCalledTimes(3);
    });
  });
});
