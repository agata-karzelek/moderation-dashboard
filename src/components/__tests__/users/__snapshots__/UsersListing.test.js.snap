// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User List Test Cases User Snapshot 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="Users"
  />
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="space-between"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      sm={4}
    >
      <Styled(MuiBox)
        pl={1}
      >
        <WithStyles(ForwardRef(Typography))
          testid="title"
          variant="h5"
        >
          Users
        </WithStyles(ForwardRef(Typography))>
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      sm={8}
    >
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
      >
        <Styled(MuiBox)
          ml={2}
        >
          <Search
            cancelSearch={[Function]}
            placeholder="Search for user name"
            requestSearch={[Function]}
            searched=""
          />
        </Styled(MuiBox)>
        <Styled(MuiBox)
          ml={2}
        >
          <FilterUsers
            isFiltered={[Function]}
          />
        </Styled(MuiBox)>
        <Styled(MuiBox)
          ml={2}
        >
          <AddUserMetadataModal
            isAdded={[Function]}
          />
        </Styled(MuiBox)>
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <br />
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={6}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <UsersTable
        data={Array []}
        deleteUser={[Function]}
        filterBy=""
        searchResult={Array []}
        setSearchableData={[Function]}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Fragment>
`;
